lib/widgets/swipable_message_row.dart

```diff
@@ -0,0 +1,83 @@
+import 'package:flutter/material.dart';
+import 'package:flutter/rendering.dart';
+
+import '../api/model/model.dart';
+import 'icons.dart';
+import 'text.dart';
+import 'theme.dart';
+
+class SwipableMessageRow extends StatefulWidget {
+  const SwipableMessageRow({
+    super.key,
+    required this.children,
+    required this.message,
+  });
+
+  final List<Widget> children;
+  final Message message;
+
+  @override
+  State<StatefulWidget> createState() => _SwipableMessageRowState();
+}
+
+class _SwipableMessageRowState extends State<SwipableMessageRow> {
+  @override
+  Widget build(BuildContext context) {
+    final hasMarker = widget.message.editState != MessageEditState.none;
+
+    return Row(
+      crossAxisAlignment: CrossAxisAlignment.baseline,
+      textBaseline: localizedTextBaseline(context),
+      children: [
+        hasMarker
+          ? _EditStateMarker(editState: widget.message.editState)
+          : const SizedBox(width: 16),
+        ...widget.children,
+      ],
+    );
+  }
+}
+
+class _EditStateMarker extends StatelessWidget {
+  /// The minimum width of the marker.
+  // Currently, only the collapsed state of the marker has been implemented,
+  // where only the marker icon, not the marker text, is visible.
+  static const double widthCollapsed = 16;
+
+  const _EditStateMarker({
+    required MessageEditState editState,
+  }) : _editState = editState;
+
+  final MessageEditState _editState;
+
+  @override
+  Widget build(BuildContext context) {
+    final designVariables = DesignVariables.of(context);
+
+    final IconData icon;
+    final double iconSize;
+
+    switch (_editState) {
+      case MessageEditState.none:
+        return const SizedBox(width: widthCollapsed);
+      case MessageEditState.edited:
+        icon = ZulipIcons.edited;
+        iconSize = 14;
+        break;
+      case MessageEditState.moved:
+        icon = ZulipIcons.message_moved;
+        iconSize = 8;
+        break;
+    }
+
+    return ConstrainedBox(
+      constraints: const BoxConstraints(maxWidth: widthCollapsed),
+      child: Padding(
+        padding: const EdgeInsetsDirectional.only(start: 5, end: 3),
+        child: OverflowBox(
+          fit: OverflowBoxFit.deferToChild,
+          maxWidth: 8,
```

Comment:
Can we do this without OverflowBox?

There's no real overflow happening in the end, right? The wider icon, for moved, is only width 9 — so it should fit just fine in the 16px width that this Padding widget has. Can we adjust the padding to match that instead?

Reply:
That's doable. I think `OverflowBo`x is something that we might need for the animated version.

Reply:
OK. Even if we end up needing it for animated, I'd like to try starting without it as a baseline - I think that'll make this easier to reason about.